drivers/imx-enet.c

```diff
@@ -0,0 +1,1435 @@
+{
+	if (timeout != 0) {
+		time_t now, when;
+		gettime(&now, NULL);
+		when = now + timeout;
+
+		while ((state->mmio->EIR & ENET_IRQ_MII) == 0) {
+			gettime(&now, NULL);
+			if (now >= when) {
+				enet_debug_printf(state, "enet_mdioWait: timeout");
+				break;
+			}
+			usleep(10);
+		}
+	}
+	else {
+		while ((state->mmio->EIR & ENET_IRQ_MII) == 0)
+			/* relax */;
+	}
+
+	state->mmio->EIR = ENET_IRQ_MII;
+}
+
+
+static uint16_t enet_mdioIO(enet_state_t *state, unsigned addr, unsigned reg, unsigned val, unsigned op)
+{
+#if MDIO_DEBUG
+	/* clang-format off */
+	enet_printf(state, "mdio: op %s, addr=0x%08x, reg=0x%08x, val=0x%04x",
+		op == ENET_MMFR_OP_READ_VAL  ? "READ"  :
+		op == ENET_MMFR_OP_WRITE_VAL ? "WRITE" :
+									  "WARN: undefined",
+		addr, reg, val);
+	/* clang-format on */
+#endif
+
+#if !ENET_MDC_ALWAYS_ON
+	state->mmio->MSCR = state->mscr;
+#endif
+
+	uint32_t mmfr = 0;
+
+	/* clause 45 */
+	if ((addr & NETDEV_MDIO_CLAUSE45) != 0) {
+		/* ST */
+		mmfr |= ENET_MMFR_ST_CLAUSE45_VAL;
+
+		uint32_t dev = ((addr & NETDEV_MDIO_A_MASK) << 18) |
+				((addr & NETDEV_MDIO_B_MASK) << (23 - 8));
+		mmfr = (ENET_MMFR_OP_ADDR << ENET_MMFR_OP_SHIFT) | /* extended MDIO data r/w */
+				(ENET_MMFR_TA_VAL << ENET_MMFR_TA_SHIFT) | dev | (reg & 0xFFFF);
+
+		state->mmio->MMFR = mmfr;
+		enet_mdioWait(state, MDIO_TIMEOUT);
+
+		mmfr = (op << ENET_MMFR_OP_SHIFT) | dev | /* standard MDIO data r/w */
+				ENET_MMFR_TA_VAL << ENET_MMFR_TA_SHIFT | ((op == ENET_MMFR_OP_READ_VAL) ? 0 : val & ENET_MMFR_DATA_MASK);
+	}
+	/* clause 22 */
+	else {
+		mmfr = (ENET_MMFR_ST_CLAUSE22_VAL << ENET_MMFR_ST_SHIFT) |                          /* ST */
+				((op << ENET_MMFR_OP_SHIFT) & ENET_MMFR_OP_MASK) |                          /* OP */
+				(((addr & NETDEV_MDIO_A_MASK) << ENET_MMFR_PA_SHIFT) & ENET_MMFR_PA_MASK) | /* PA - PHY addr */
+				(((reg & NETDEV_MDIO_A_MASK) << ENET_MMFR_RA_SHIFT) & ENET_MMFR_RA_MASK) |  /* RA - reg addr */
+				ENET_MMFR_TA_VAL << ENET_MMFR_TA_SHIFT |                                    /* TA */
+				((op == ENET_MMFR_OP_READ_VAL) ? 0 : val & ENET_MMFR_DATA_MASK);            /* DATA */
+	}
+
+	state->mmio->MMFR = mmfr;
+	enet_mdioWait(state, MDIO_TIMEOUT);
+
+	val = state->mmio->MMFR & ENET_MMFR_DATA_MASK;
+
+#if !ENET_MDC_ALWAYS_ON
+	state->mmio->MSCR = 0;
+#endif
+
+	return val;
+}
+
+
+static uint16_t enet_mdioRead(void *arg, unsigned addr, uint16_t reg)
+{
+	enet_state_t *state = arg;
+	return enet_mdioIO(state, addr, reg, 0, ENET_MMFR_OP_READ_VAL);
+}
+
+
+static void enet_mdioWrite(void *arg, unsigned addr, uint16_t reg, uint16_t val)
+{
+	enet_state_t *state = arg;
+	(void)enet_mdioIO(state, addr, reg, val, ENET_MMFR_OP_WRITE_VAL);
+}
+
+
+static const mdio_bus_ops_t enet_mdio_ops = {
+	enet_mdioSetup,
+	enet_mdioRead,
+	enet_mdioWrite,
+};
+
+
+static int platformctl_seq(const platformctl_t *pctl, size_t n)
+{
+	int err = 0;
+
+	for (; n--; ++pctl) {
+		err = platformctl((platformctl_t *)pctl);
```

Comment:
This leads to undefined behavior. Please declare a local `platformctl_t` variable and copy the input into it.